Diff improvements

silverwind opened the pull request:
- Avoid flash of wrong tree toggle icon by setting icon state synchronously
- Avoid "pop-in" of tree by leaving space based on sync state
- Use the same border/box-shadow combo used on comment `:target` also for file `:target`.

<img width="1250" alt="Screenshot 2023-03-18 at 00 52 04" src="https://user-images.githubusercontent.com/115237/226071392-6789a644-aead-4756-a77e-aba3642150a0.png">
<img width="1246" alt="Screenshot 2023-03-18 at 00 59 43" src="https://user-images.githubusercontent.com/115237/226071443-8bcba924-458b-48bd-b2f0-0de59cb180ac.png">

(Yes I'm aware the border-radius in bottom corners is suboptimal, but this would be notorously hard to fix without relying on `overflow: hidden`).

silverwind commented:
I would say this should go into 1.19.0 release, because the backport https://github.com/go-gitea/gitea/pull/23546 includes the header regression.

silverwind commented on web_src/css/repository.css:
```diff
@@ -1846,6 +1846,24 @@
+
+.diff-file-box .ui.attached.header,
+.diff-file-box .ui.attached.table {
+  margin: 0; /* remove fomantic negative margins */;
+  width: initial; /* remove fomantic over 100% width */;
+  max-width: initial; /* remove fomantic over 100% width */;
```
```suggestion
  margin: 0; /* remove fomantic negative margins */
  width: initial; /* remove fomantic over 100% width */
  max-width: initial; /* remove fomantic over 100% width */
```

silverwind commented on web_src/css/repository.css:
```diff
@@ -1846,6 +1846,24 @@
+  margin: 0; /* remove fomantic negative margins */
+  width: initial; /* remove fomantic over 100% width */
+  max-width: initial; /* remove fomantic over 100% width */
+}
```
Should try later to make these styles global. Fomantic's use of -1 left/right margin and 100% + 2px width on these `attached` fomantic elements makes certain effects like border impossible because it makes elements break outside their container.

silverwind commented on web_src/css/repository.css:
```diff
@@ -1846,6 +1846,24 @@
+  border: 1px solid transparent;
+  border-radius: 0.285rem; /* Just like ui.top.attached.header */
+}
+
```
```suggestion

/* TODO: try making these rules global to fix elements breaking out of their parent box */
```

delvh commented on web_src/css/repository.css:
```diff
@@ -1846,6 +1846,25 @@
+.diff-file-box:target {
+  border-color: var(--color-primary) !important;
+  border-radius: var(--border-radius) !important;
+  box-shadow: 0 0 0 3px var(--color-primary-alpha-30) !important;
```
Why not `--color-accent` anymore?
That's the semantically correct color here…

delvh replied:
> Use the same border/box-shadow combo used on comment :target also for file :target.

Or better: Why not convert the other color to `--color-accent` as well?

silverwind replied:
The style is the exact same as from comment highlight. The box shadow requires an alpha channel color, but without CSS [`color-mix`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) I'm afraid we can not automatically derive this alpha-channel color. So I suggest leave it like that.

silverwind replied:
BTW to me the distinction between `--color-accent` and `--color-primary` is not clearly defined. For all intends and puposes, they should be the same color imho.

silverwind replied:
I think I might even merge the rules for comment and file highlight into one so it's clear they should be the same colors.

delvh replied:
Well, for me, the difference is the following:
- `color-primary`: UI elements that are supposed to be interacted (`Ok`/`Save` buttons, …)
- `color-accent`: UI elements that should receive a highlighting, but the highlighting is purely for stylistic purposes, and it is not said, if you can interact with this element…

silverwind replied:
As per [MDN](https://developer.mozilla.org/en-US/docs/Glossary/Accent):

> An accent is a typically bright color that contrasts with the more utilitarian background and foreground colors within a color scheme.

> On the web, an accent is sometimes used in [<input>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) elements for the active portion of the control, for instance the background of a checked [checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox).

I think the use case exactly matches that of `--primary-color`, so I see no reason to make them different. Accent has nothing to do with highlight imho.

silverwind replied:
Merged the rules in https://github.com/go-gitea/gitea/pull/23553/commits/6f295f13f9ee71614821a0a17fd212521a4bcaad. I think we should leave those as `--primary-color`. When CSS `color-mix` is available, we could re-evaluate deriving alpha-channel colors off it and maybe apply it to the rule then, but currently this needs extra CSS vars which is problematic for themes.

silverwind commented on web_src/js/components/DiffFileTree.vue:
```diff
@@ -109,15 +110,20 @@ export default {
-      const diffTree = document.getElementById('diff-file-tree');
-      diffTree.classList.toggle('gt-hidden', !visible);
+      const tree = document.getElementById('diff-file-tree');
+      if (visible) {
+        showElem(tree);
+        hideElem(toShow);
+        showElem(toHide);
+      } else {
+        hideElem(tree);
+        hideElem(toHide);
+        showElem(toShow);
+      }
```
I was afraid of `toggleElem` because it had these runMode checks etc, but reading through it, it seems like that's something slated for future removal.

silverwind replied:
Done

silverwind commented:
Another small but useful change: Review comment boxes now left-align and fit the same content per line as GitHub's boxes.

<img width="1232" alt="Screenshot 2023-03-19 at 21 44 25" src="https://user-images.githubusercontent.com/115237/226207964-935708d4-fe62-41f6-b3c0-8a592d1279df.png">
<img width="1231" alt="Screenshot 2023-03-19 at 21 44 16" src="https://user-images.githubusercontent.com/115237/226207963-4a563709-e450-42c9-9ee7-ac4ea92d9e3a.png">

silverwind commented:
Ready again.

silverwind commented:
There is one more bug I'll try to fix. With many files in a diff, the tree shrinks to a tiny height. [Example page](https://try.gitea.io/silverwind/symlink-test/compare/a12218aa4eb65e802eea1cd691a0d9f0f33e306a...master). I have a fix the almost works but the tree scrolls out of view when the page end is reached, so it's not ideal.

silverwind commented:
Diff tree height fix is best-effort. It still scrolls partially out of view when page end is reached, but I see no obvious way to fix and it's only a minor inconvenience.

gempir commented:
> Avoid "pop-in" of tree on page load by leaving space based on sync state

Love this!

silverwind commented:
> Just curious, what's the difference between the old behavior and the new behavior? I guess the old behavior wont' cause "the tree scrolls partially out of view when the page end is reached"?

The 100% height does not work and is the cause for the list becoming tiny. Here is what I mean with the partial scroll-out. I think it's very minor:

![scroll](https://user-images.githubusercontent.com/115237/226377855-80b2c8b1-37e5-4c3b-bbf9-d4598823cb67.gif)

wxiaoguang commented:
> The 100% height does not work and is the cause for the list becoming tiny

Before https://github.com/go-gitea/gitea/pull/23554 , it works well?

silverwind commented:
Indeed, removing the `.diff-detail-box` explicit height also makes `height: 100%` work and fixes the scroll-out. Checking if there is a solution that accomodates all issues...

silverwind commented:
Problem is thought, the natural height of `.diff-detail-box` is dictated by font size so we really want to define it, otherwise the element will be inconsistent height depending on font...

wxiaoguang commented:
Actually, either approach could look good to me (I can accept either, it doesn't bother me 😁 ). But I guess the old behavior is done by purpose, if it's too difficult to co-work with other elements, maybe it's not a must.

If we'd like to use the new behavior, maybe we can leave a comment for future developers.

silverwind commented:
Found a solution that satisfies all cases. Setting the action button to fixed height makes the stickied element basically fixed height at 47/77px and the height: 100% also works with it on the tree, preventing scroll-out:

<img width="484" alt="Screenshot 2023-03-20 at 16 04 58" src="https://user-images.githubusercontent.com/115237/226382438-79a1bd96-e7bf-444d-9037-f3ab0614ce5a.png">

silverwind commented:
I pushed `b9e8595e53` to this branch but GitHub UI does not show it for some reason.

silverwind commented:
[Here is the commit](https://github.com/silverwind/gitea/commit/b9e8595e53cbf0ed855961a179310286ffb18c63). I hope I didn't break this PR 😉

silverwind commented:
Okay, seems GitHub UI finally caught up. PR Should be good to go now, please review.

HesterG commented:
This is really confusing to me, that in one of my repo, the `diff-file-tree` has a computed width of 0, but I am not sure why this happened.. And I don't know if you have encountered this.

https://user-images.githubusercontent.com/17645053/226502086-e38d3dc7-bc75-43a2-90f2-8ee9809b4b6f.mov

If this is an issue, adding percentage width to `diff-file-tree`, this might fix this as shown in the end of the video above.

HesterG commented:
> This is really confusing to me, that in one of my repo, the `diff-file-tree` has a computed width of 0, but I am not sure why this happened.. And I don't know if you have encountered this.
> 
> If this is an issue, adding percentage width to `diff-file-tree` might fix this as shown in the end of the video above.

I created a new gitea from build from latest `main` branch, and this doesn't happen again, so might ignore this..

silverwind commented:
I've changed the toggle `<a>` to an `<button>` as it should be, and added a few simple button styles for it to the base rules. Much better for accessibility.

I think the only thing missing from that button is a tooltip, will check.

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -15,11 +15,17 @@
+					{{svg "octicon-sidebar-expand" 20 "icon gt-hidden"}}
+				</button>
+				<script>
+					const treeVisible = localStorage?.getItem('diff_file_tree_visible') === 'true';
+					const btn = document.querySelector('.diff-toggle-file-tree-button');
+					const icon = `.octicon-sidebar-${treeVisible ? 'expand' : 'collapse'}`;
```
These lines will case: `window.btn`, `window.icon`, etc

silverwind replied:
How to avoid? I can't set `type=module` unfortunately as I need synchronous execution.

silverwind replied:
Does not seem to be the case at least in Firefox:

<img width="172" alt="image" src="https://user-images.githubusercontent.com/115237/226631855-10ce9efa-10db-49f1-87ec-c90eeb98d3b1.png">

wxiaoguang replied:
Maybe comment + prefixWithLongEnoughName to avoid conflicts, or CustomElements.

In history, various kinds of pollution already caused a lot of problems.

silverwind replied:
CustomElements does not work easily in this case as I need access to the svg data, which means I'd have to import `svg.js` into the js chunk which would considerably blow it up in size and duplicate the SVG data into it.

silverwind replied:
Added variable prefixes.

silverwind commented on web_src/css/base.css:
```diff
@@ -238,6 +238,11 @@ table {
   border-collapse: collapse;
 }
 
+button {
+  cursor: pointer;
+  color: var(--color-text);
+}
```
For some reason my button remained at cursor `pointer` and this rule also matches GitHub's CSS which has it.

wxiaoguang replied:
If we decide to make all buttons have `cursor: pointer;`, then that's fine.

silverwind commented:
Tooltip added:

![](https://user-images.githubusercontent.com/115237/226631329-01532434-c8ec-4c40-af15-e3a2553ee877.gif)

wxiaoguang commented on web_src/js/components/DiffFileTree.vue:
```diff
@@ -113,8 +113,12 @@ export default {
       const tree = document.getElementById('diff-file-tree');
+      const newTooltip = btn.getAttribute(`data-content-${visible ? 'hide' : 'show'}`);
+      btn.setAttribute('data-content', newTooltip);
+      btn._tippy.setContent(newTooltip);
```
I don't think this code is safe. 

There is no init order constraint, there might be no `btn._tippy` because the "tooltip" module might not be inited?

silverwind replied:
It does run on component mount, so there is a risk. I'll see to not trying to update at mount as the tooltip will already be correct there from the HTML script.

wxiaoguang replied:
How about https://github.com/go-gitea/gitea/pull/23574 ?  With that PR, developers only need to change "data-tooltip-content", they do not need to call `btn._tippy.setContent` manually.

silverwind replied:
Updated it so it skips the tooltip update on mount.

https://github.com/go-gitea/gitea/pull/23574 with the MutationObserver does sound good, but I can't leave out this update now, so we need to adjust PRs depending on which one lands first.

silverwind replied:
https://github.com/go-gitea/gitea/pull/23574 is merged, so let's see if I can skip the `setContent` now.

silverwind replied:
Does not appear to work. The `data-content` attribute is updated, but the tooltip does not change.

silverwind replied:
Yeah, works now after update to `data-tooltip-content`.

wxiaoguang commented:
A little off-topic, this PR has became larger and larger, and it contains some uncertain parts (for example, I would propose to use a general tooltip solution, instead of patching more code to `updateState`.


And since 1.19 is waiting for the bug fix, should we fix the bug and backport first, then fine tune more?

silverwind commented:
We could extract just the bugfix parts into another PR.

silverwind commented:
Extract done in https://github.com/go-gitea/gitea/pull/23616.

silverwind commented:
Just rebasing main branch onto this, I notice two new isses introduced by recent commits:

1. File tree items have lost color and are misaligned
2. Tooltip does not show, e.g. fallback to `data-content` does not work.

<img width="245" alt="image" src="https://user-images.githubusercontent.com/115237/227734799-3fd7674d-3ee0-46a9-838e-f0268c07fc48.png">

silverwind commented:
Both issues fixed, they were regressions from https://github.com/go-gitea/gitea/pull/23570.

wxiaoguang commented on templates/repo/diff/box.tmpl:
```diff
@@ -15,11 +15,18 @@
 		<div class="diff-detail-box diff-box sticky gt-df gt-sb gt-ac gt-fw">
 			<div class="gt-df gt-ac gt-fw">
-				<a class="diff-toggle-file-tree-button muted gt-df gt-ac">
+				<button class="diff-toggle-file-tree-button tooltip gt-df gt-ac" data-show-text="{{.locale.Tr "repo.diff.show_file_tree"}}" data-hide-text="{{.locale.Tr "repo.diff.hide_file_tree"}}">
```
Remove the "tooltip" class?

silverwind replied:
Ah, it was already made obsolete, I missed that.

silverwind replied:
Done

wxiaoguang commented on web_src/js/components/DiffFileTreeItem.vue:
```diff
@@ -1,7 +1,7 @@
   <div v-show="show" :title="item.name">
     <!--title instead of tooltip above as the tooltip needs too much work with the current methods, i.e. not being loaded or staying open for "too long"-->
-    <div class="item" :class="item.isFile ? 'filewrapper gt-p-1' : ''">
+    <div class="item" :class="item.isFile ? 'filewrapper gt-p-1 gt-ac' : ''">
```
I guess these styles could be put into `filewrapper` class?

silverwind replied:
See below

wxiaoguang commented on web_src/js/components/DiffFileTreeItem.vue:
```diff
@@ -20,7 +20,7 @@
 
       <!-- Directories -->
-      <div v-if="!item.isFile" class="directory gt-p-1" @click.stop="handleClick(item.isFile)">
+      <div v-if="!item.isFile" class="directory gt-p-1 gt-ac" @click.stop="handleClick(item.isFile)">
```
(as above)

silverwind replied:
Can not put into `.filewrapper` CSS because directories have no hover effect, only files.

silverwind replied:
But I guess I can do a bit of refactoring here anyways...

silverwind replied:
Actually, no, I don't think it's worth the effort trying to do this with CSS. It's be a net increase in code and would likely require a new classname or new CSS rule that matches both.

wxiaoguang commented on web_src/js/components/DiffFileTree.vue:
```diff
@@ -109,15 +108,17 @@ export default {
+      const btn = document.querySelector('.diff-toggle-file-tree-button');
+      const [toShow, toHide] = btn.querySelectorAll('.icon');
+      const tree = document.getElementById('diff-file-tree');
+      const newTooltip = btn.getAttribute(`data-${visible ? 'hide' : 'show'}-text`);
```
To make the attribute searchable, how about:

```suggestion
      const newTooltip = btn.getAttribute(visible ? 'data-hide-text' : 'data-show-text');
```

I think we have done so when accessing `data-foo-bar`: avoid using `.dataset.fooBar` or `.data('foo-bar')`.

silverwind replied:
Yeah we can do this for easier greppability. Also need to do the same change in JS.

HesterG commented on web_src/js/components/DiffFileTree.vue:
```diff
@@ -109,15 +108,17 @@ export default {
+      const tree = document.getElementById('diff-file-tree');
+      const newTooltip = btn.getAttribute(visible ? 'data-hide-text' : 'data-show-text');
+      btn.setAttribute('data-tooltip-content', newTooltip);
+      toggleElem(tree, visible);
```
For  `DiffFileTree.vue`, looks like the `v-if="fileTreeIsVisible"` is not necessary? Because the visibility is toggled by the logic here as well as in script of `box.tmpl`:
```
<script>
	if (diffTreeVisible) document.getElementById('diff-file-tree').classList.remove('gt-hidden');
</script>
```

silverwind replied:
The `<script>` only runs once on page load to immediately set the correct state after the element is rendered into the DOM. This is to prevent flashes of incorrect content.

The vue logic is still needed to achieve toggle behaviour. I think the only tiny inefficiency is that we read `localStorage` twice, but I'd argue it's probably not worth the effort trying to query current state from the DOM.

HesterG replied:
You are right, I ignored the fact that the script is a one time running thing.
